File: 528-swapping-nodes-in-a-linked-list/swapping-nodes-in-a-linked-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* kthFromBegining(ListNode* head, int k){
        while(k > 0 && head){
            head = head->next; 
            k --; 
        }
        return head; 
    }
    ListNode* kthFromEnd(ListNode* head, int k){
        ListNode* tmp = head; 
        while(k>0){
            tmp = tmp->next;
            k --;
        }
        while(tmp->next){
            tmp = tmp->next; 
            head = head->next; 
        }
        return head; 
    }
    ListNode* swapNodes(ListNode* head, int k) {
       if(k == 0 ) return head; 
       ListNode* tmp1 = head; 
       ListNode* tmp2 = head;
       tmp2 = kthFromBegining(tmp2, k-1);
       tmp1 = kthFromEnd(tmp1, k-1);
       swap(tmp1->val, tmp2->val); 
       return head; 
    };
};

File: 528-swapping-nodes-in-a-linked-list/swapping-nodes-in-a-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def swapNodes(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:

        if not head : 
            return None

        # k = 2
        # swapping nodes in linked list. 
        node = head
        lst = head  
        tmp = head 

        while k > 1:
            lst = lst.next 
            k-=1

        kth_from_start = lst  

        while lst.next :
            lst = lst.next
            tmp = tmp.next 
        
        kth_from_end = tmp
        
        kth_from_start.val , kth_from_end.val = kth_from_end.val, kth_from_start.val 

        return head 

                   
            

        